wagtail/core/blocks/stream_block.py

```diff
@@ -543,6 +543,33 @@ def get_prep_value(self):
 
         return prep_value
 
+    def blocks_by_type(self):
+        blocks = {}
+        for item in self:
+            block_type = item.block_type
+            block_value = item.value
+
+            if block_type in blocks:
+                blocks[block_type].append(block_value)
+            else:
+                blocks[block_type] = [block_value]
+
+        return blocks
+
+    def blocks_of_type(self, block_type):
+        if block_type not in self.stream_block.child_blocks:
```

Comment:
I don't think this test is quite right: if `block_type` is not a recognised block type for the StreamField (as in your `invalid_type` test), it will do the right thing, but if `block_type` _is_ a recognised type but there are no instances of it in the stream, then it will pass this test and `self.blocks_by_type()[block_type]` will still fail with a KeyError.

Reply:
Ok, so should I do a try/except and if we catch a `KeyError` we return None or an empty list depending on the case?

Reply:
Yes, something like that. I don't think we need to distinguish between the "unrecognised block type" and "no instances found" cases - we can just return the 'empty' result (None or empty list). So I think we can leave out the `if block_type not in self.stream_block.child_blocks` check entirely and just do something like: `return self.blocks_by_type().get(block_type, [])`

Reply:
Now that you've updated the `return` line, this check can be removed completely.